### ISSUE TITLE
Instruction set compilation flags should always be enabled for SIMD implementations

### ISSUE THREAD

**Issue by zorun** [weidai11/cryptopp#576]
I am building crypto++ 6.0.0 for Arch Linux, and noticed that the new SIMD files are *not* built with the needed flags (`-msse4.1`, `-maes`, etc) by default.

These flags should be enabled inconditionnally:

- these flags are necessary (otherwise, build errors occur, like mentioned in the release notes or #573 );
- the needed CPU features are already checked at run-time (if I read `cpu.cpp` correctly) so there's no harm enabling all features at build-time;
- when packaging for a distro, the build-time and run-time CPU have different capabilities, so it makes no sense to test CPU features at build time.

Note that I am using the `GNUMakefile-cross` build system, because the normal `GNUMakefile` is not suitable for packaging (it sets various unwanted build flags with no simple way to disable them).

What I propose is something like this:

```
ifneq ($(IS_i686)$(IS_x86_64),00)
  ARIA_FLAG = -mssse3
  BLAKE2_FLAG = -msse4.2
  AES_FLAG = -msse4.1 -maes
  # etc
endif
```
that is, remove all feature detection like `$(shell echo | $(CXX) -x c++ $(CXXFLAGS) -mssse3 -mpclmul -dM -E - 2>/dev/null | $(EGREP) -i -c __PCLMUL__ )`

I tested doing just that and building on an old Core2 (which certainly doesn't have AES, SHA and so on).  It builds fine and all tests pass, printing the following:

```
hasSSE2 == 1, hasSSSE3 == 1, hasSSE4.1 == 1, hasSSE4.2 == 0, hasAESNI == 0, hasCLMUL == 0, hasRDRAND == 0, hasRDSEED == 0, hasSHA == 0, isP4 == 0
```
so runtime feature detection seems to work fine.

**Comment by noloader** [comment]
Thanks @zorun.

> I am building crypto++ 6.0.0 for Arch Linux, and noticed that the new SIMD files are not built with the needed flags (-msse4.1, -maes, etc) by default.

Hmm... I'm seeing the flags added to the SIMD files on my Arch Linux version 233 test machine.

-----

> Note that I am using the GNUMakefile-cross build system, because the normal GNUMakefile is not suitable for packaging

OK, you should be using `GNUmakefile`. `GNUmakefile` works as expected on my Arch64 machine.

`GNUmakefile-cross` is for Android, some ARM and iOS.

-----

> What I propose is something like this ...

Yeah, that does not work on older machines like CentOS 5 or Red Hat 6. It does not work on newer machines with older assemblers and linkers, like OpenBSD 6.

**Comment by noloader** [comment]
@zorun,

I'm going to close this since we merged your PR.

I recommend you use `GNUmakefile`, and not `GNUmakefile-cross`. You are certainly free to use procedures that you prefer. But keep in mind we might not be able to support all of your use cases, and you may have to develop your own set of patches.

**Comment by zorun** [comment]
@noloader indeed, the regular `GNUMakefile` works fine, thanks!

I remember that at some point it did not fully honor `CXXFLAGS` and was adding unwanted flags (like `-g` and random `-march`).  But it seems to be fine now.

In any case, I think I misunderstood the checks I was suggesting to remove.  They do not seem to depend on CPU features but rather on compiler capabilities.  Sorry for the incorrect report.

**Comment by noloader** [comment]
> I remember that at some point it did not fully honor CXXFLAGS and was adding unwanted flags (like -g and random -march).

Yes, if you don't do anything special, then we add flags for the typical use case. For us, the typical use case is the average user. They get flags like `-DNDEBUG`, `-g2`, `-O3` and `-fPIC`.

In the past, before we switched to BASE+SIMD, I think we had flag problems. If a distro removed `-march=native` then we often needed to disable advanced features like SSE3, SSE4, AES and SHA. That sucked for distros and their users, and it was fixed at 6.0 when we switched to BASE+SIMD.
